atuin-client/src/database.rs

```diff
@@ -286,27 +287,96 @@ impl Database for Sqlite {
         let query = query.to_string().replace('*', "%"); // allow wildcard char
         let limit = limit.map_or("".to_owned(), |l| format!("limit {}", l));
 
-        let query = match search_mode {
-            SearchMode::Prefix => query,
-            SearchMode::FullText => format!("%{}", query),
-            SearchMode::Fuzzy => query.split("").join("%"),
+        let (query_sql, query_params) = match search_mode {
+            SearchMode::Prefix => ("command like ?1".to_string(), vec![format!("{}%", query)]),
+            SearchMode::FullText => ("command like ?1".to_string(), vec![format!("%{}%", query)]),
+            SearchMode::Fuzzy => {
+                let split_regex = Regex::new(r" +").unwrap();
+                let terms: Vec<&str> = split_regex.split(query.as_str()).collect();
+                let num_terms = terms.len();
+                let mut query_sql = "".to_string();
+                let mut query_params = std::vec::Vec::with_capacity(num_terms);
+                let mut was_or = false;
+                for (i, query_part) in terms.into_iter().enumerate() {
+                    // TODO smart case mode could be made configurable like in fzf
+                    let (operator, glob) = if query_part.contains(char::is_uppercase) {
+                        ("glob", '*')
+                    } else {
+                        ("like", '%')
+                    };
+                    let (is_inverse, query_part) = if query_part.starts_with('!') {
+                        (true, query_part.strip_prefix('!').unwrap())
+                    } else {
+                        (false, query_part)
+                    };
+                    match query_part {
+                        "|" => {
+                            if !was_or {
+                                query_sql.push_str(" OR ");
+                                was_or = true;
+                                continue;
+                            } else {
+                                query_params.push(format!("{glob}|{glob}", glob = glob));
+                            }
+                        }
+                        exact_prefix if query_part.starts_with('^') => query_params.push(format!(
+                            "{term}{glob}",
+                            term = exact_prefix.strip_prefix('^').unwrap(),
+                            glob = glob
```

Comment:
Minor nit: this `glob = glob` syntax is unnecessary on the latest version ([since 1.58.0](https://blog.rust-lang.org/2022/01/13/Rust-1.58.0.html#captured-identifiers-in-format-strings))

Reply:
removed the `glob = glob`